backends/simplego/dotgeneral_fastpath.go

```diff
@@ -0,0 +1,283 @@
+//   3. Batched MatMul: [B, M, K] × [B, K, N] → [B, M, N] (contract on lhs axis 2, rhs axis 1)
+//   4. Multi-batch: [B1, B2, M, K] × [B1, B2, K, N] → [B1, B2, M, N]
+//
+// See also: execDotGeneralSmallNormalized which transposes to [Batch, Cross, Contract] form
+// where BOTH operands have the contracting dimension last (sequential access for both).
+func isContractLastOrder(lhsShape, rhsShape shapes.Shape, lhsContractingAxes, rhsContractingAxes, lhsBatchAxes, rhsBatchAxes []int) bool {
+	lhsRank := lhsShape.Rank()
+	rhsRank := rhsShape.Rank()
+
+	// Check for standard matrix-matrix multiplication: [M, K] × [K, N]
+	if lhsRank == 2 && rhsRank == 2 &&
+		len(lhsContractingAxes) == 1 && len(rhsContractingAxes) == 1 &&
+		len(lhsBatchAxes) == 0 && len(rhsBatchAxes) == 0 {
+		// Contracting: lhs last axis (1) with rhs first axis (0)
+		if lhsContractingAxes[0] == 1 && rhsContractingAxes[0] == 0 {
+			return true
+		}
+	}
+
+	// Check for matrix-vector multiplication: [M, K] × [K]
+	if lhsRank == 2 && rhsRank == 1 &&
+		len(lhsContractingAxes) == 1 && len(rhsContractingAxes) == 1 &&
+		len(lhsBatchAxes) == 0 && len(rhsBatchAxes) == 0 {
+		// Contracting: lhs last axis (1) with rhs only axis (0)
+		if lhsContractingAxes[0] == 1 && rhsContractingAxes[0] == 0 {
+			return true
+		}
+	}
+
+	// Check for batched matrix multiplication: [B, M, K] × [B, K, N]
+	if lhsRank == 3 && rhsRank == 3 &&
+		len(lhsContractingAxes) == 1 && len(rhsContractingAxes) == 1 &&
+		len(lhsBatchAxes) == 1 && len(rhsBatchAxes) == 1 {
+		// Batch on first axis (0)
+		if lhsBatchAxes[0] == 0 && rhsBatchAxes[0] == 0 {
+			// Contracting: lhs axis 2 with rhs axis 1
+			if lhsContractingAxes[0] == 2 && rhsContractingAxes[0] == 1 {
+				return true
+			}
+		}
+	}
+
+	// Check for multi-batch matmul: [B1, B2, M, K] × [B1, B2, K, N]
+	if lhsRank == 4 && rhsRank == 4 &&
+		len(lhsContractingAxes) == 1 && len(rhsContractingAxes) == 1 &&
+		len(lhsBatchAxes) == 2 && len(rhsBatchAxes) == 2 {
+		// Batch on first two axes
+		if lhsBatchAxes[0] == 0 && lhsBatchAxes[1] == 1 &&
+			rhsBatchAxes[0] == 0 && rhsBatchAxes[1] == 1 {
+			// Contracting: lhs axis 3 with rhs axis 2
+			if lhsContractingAxes[0] == 3 && rhsContractingAxes[0] == 2 {
+				return true
+			}
+		}
+	}
+
+	return false
+}
+
+// isBatchCrossContractOrder checks if the tensor axes are ordered as:
+// [Batch..., Cross..., Contract...]
+//
+// This ordering means:
+//   - Batch axes come first (axis 0, 1, ...)
+//   - Cross axes come next (the "output" dimensions)
+//   - Contracting axes come last (enabling sequential memory access for dot products)
+//
+// When axes are in this order, no transpose is needed before the normalized
+// DotGeneral computation.
+func isBatchCrossContractOrder(shape shapes.Shape, contractingAxes, batchAxes []int) bool {
+	rank := shape.Rank()
+	if rank == 0 {
+		return true
+	}
+
+	// Build expected order: batch axes first, then cross, then contracting
```

Comment:
Maybe a simpler way is checking that:

- Batch axes come first: 
  `len(batchAxes) == 0 || (max(batchAxes) == len(batchAxes)-1` 
  
- Contracting axes come last:  
   `len(contractingAxes) == 0 || min(contractingAxes) == shape.Rank() - len(contractingAxes)`

Notice that the DotGeneral already checks (I think) that batchAxes and contractingAxes are unique values.